### DISCUSSION TITLE
Support for custom field annotations

### DISCUSSION THREAD

**Discussion by rjpipino** [instancio/instancio#913] [Q&A]
I haven't found a good way to handle data generation for custom bean validators.  Suppose I have a custom annotation `@WithKeys` that checks a map has specific keys defined. If I have a field that looks like this:

```
@NotNull
@Size(min = 2, max = 2)
@WithKeys({"foo","bar"})
private Map<String, @NotNull @Positive Integer> map;
```

In order for the map to pass the custom validator it must contain two keys: `foo` and `bar`. If we were to implement `InstancioServiceProvider` and set the keys

```
var withKeys = field.getAnnotation(WithKeys.class);
if (withKeys != null) {
    return generators.map().withKeys(withKeys.value());
}
 ``` 
 
This will work _except_ the map generated **may not** be valid for the other validation annotations. For example, `@Size` is ignored and map generated will contain 3 entries. I think this is a **bug** and the other constraint annotations should be honored.

I could see a use case for supporting a post process on field generation. In this example, would be nice to have a way to provide a field predicate that determines if the field is annotated with the custom annotation and then supply a `BiConsumer` where the inputs are the generated instance and the field itself.

```
var foo = Instancio.of(Foo.class)
    .onFieldComplete((field) -> field.isAnnotationPresent(WithKeys.class),  (field, obj) -> {
        var requiredKeys = field.getAnnotation(WithKeys).value();
         // logic to set required keys in the already generated valid map
     }).create();
```                 

I think this is similar to the idea with `onComplete` except the callback needs to execute on the field being generated.

Anyway, hopefully this illustrates the problem I am trying solve and looking forward to some discussion on the idea.  If there is a way to accomplish this with the current API please provide an example 😃 

Thanks!

**Comment by armandino** [comment]
Currently there's no way to hook into bean validation processing to handle custom annotations like `@WithKeys` in your example. I'll need to see how feasible it is to implement and what the API might look like.

I see your point regarding generators returned by the SPI not honouring bean validation annotations. This behaviour is intentional. The library does not modify user-provided generators. This is to avoid unwanted behaviour. For example, if you were to specify `generators.map().size(1)`, the generated map would be of size 1 rather than the size from the annotation.

I think in this case, your best bet might be to get the `@Size` annotation along with `@WithKeys` and set the parameters in the provider:

```java
generators.map().withKeys(withKeys.value()).minSize(...).maxSize(...);
```

It is unfortunately a bit tedious :)

Regarding using `BiConsumer` in the callback, that's an interesting idea. I'll need to experiment a bit with this one.

**Comment by armandino** [comment]
@rjpipino processing custom annotations will be supported in the next release. If you have any feedback, please let me know.

- #975

**Reply by rjpipino** [reply]
Looks great! Thank you!

**Reply by armandino** [reply]
I published `4.5.0` containing the new API. Let me know if you run into any issues.
